lib/spack/spack/config.py

```diff
@@ -889,7 +889,10 @@ def add(fullpath, scope=None):
             break
 
     if has_existing_value:
-        path, _, value = fullpath.rpartition(":")
+        path, value, _ = fullpath.rpartition(components[-1])
+        # Breaking down path in this way leaves a dangling
+        # separator, remove it
+        path, *_ = path.rpartition(":")
```

Comment:
I don't think the dangling colon is crucial: Your test for example will result in `path = "config:install_tree:root"` (no last colon). The approach I suggested will replace `::` with `:` in the middle of the path, but that isn't necessary for access (which is what is being done in this context).

Are you saying that when you try `spack config add config:install_tree:root::"example"` that it fails unless the path is `config:install_tree:root:`?

Whatever happens, it doesn't make sense to set `value` using rpartition, it is just `components[-1]`. I'm also surprised it works to `rpartition` with `components[-1]` at all since that omits the quotes (e.g. you'd be left with `example:config:path:"` as the string). I'm inclined to say that it would be preferable to avoid it for `path` as well, but if I'm overlooking some case where an intermediate `::` is required for access, I can see it having more appeal. In that case though, I'd rather `process_config_path` got updated to return this preserved string (that function should handle breaking things up into sensible parts and users of that function shouldn't have to worry about reassembling them).

Reply:
The dangling `:` is necessary for overrides. The `install_tree` example doesn't use overrides, but the test I referenced does. If I remove the dangling colon, all override related tests that pass through the `add` interface fail because we're not processing them as overrides due to the dropped `:`. 
In the docstring of `process_config_path` : 
`Process a path argument to config.set() that may contain overrides ('::' or trailing ':')`
Where the overrides are denoted by the trailing `:`. It is an expected and established part of that interface. 

Sure, we don't need to use `rpartition` to set value, I was more just taking advantage of a convenient side effect there, very much not a crucial piece of code. 
We can just go with my alternative suggestion if we want to avoid `rpartition` altogether. 

I disagree that this should ever be the responsibility of `process_config_path`. As you said, that method is responsible for breaking down the "paths" into sensible component parts, and as such, should not be responsible for reconstructing them re: single responsibility. If the method is responsible for breaking things down, the caller should reasonably expect to have to reconstruct things if they need things reconstructed. 
Additionally, the caller already has the original string, they called the method with it, and reconstructing based on some arbitrary heuristic in `process_config_path` doesn't really seem globally useful to callers as opposed to allowing them reconstruct their own strings based on contextual need. 
My example also doesn't reconstruct the return of `process_config_path` it uses the original string to understand the path outside the context of the value, which `process_config_path` successfully decomposes into something clear and accessible.

Reply:
The following diff applied to the tip of this PR avoids the use of rpartition:

```
diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
index 79e33fc21a..b0eb52d019 100644
--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -888,12 +888,11 @@ def add(fullpath, scope=None):
                 value = {component: value}
             break
 
+    if override:
+        path += "::"
+
     if has_existing_value:
-        path, value, _ = fullpath.rpartition(components[-1])
-        # Breaking down path in this way leaves a dangling
-        # separator, remove it
-        path, *_ = path.rpartition(":")
-        value = syaml.load_config(value)
+        value = syaml.load_config(components[-1])
         existing = get(path, scope=scope)
```

and appears to pass the config tests. I noticed there is already a reassembly of `path` above. I think in fact 

`value = syaml.load_config(components[-1])`

could be moved above the loop, which might further clarify things. There's a couple instances of variable shadowing which make this harder to reason about.

Reply:
That does appear to ensure all tests pass, I'll apply it.

Out of curiosity, what is the issue with:
```path = ":".join(fullpath.split(":")[:-1])
value = components[-1]
```
Which sidesteps the need for logic in the first place?

Reply:
> Out of curiosity, what is the issue with:

```
path = ":".join(fullpath.split(":")[:-1])
value = components[-1]
```

I thought at the end of https://github.com/spack/spack/pull/39279#discussion_r1310378871 you were saying that was less obvious than rpartition so:

* I thought rpartition was too complex
* You offered that solution, which to you was also more complex
* I offered another possible solution, hoping it was less complex than the 2 prior options

Perhaps you think this suggestion is also strictly worse? I'm open to other possibilities.

Reply:
Oh, no, I certainly don't have a better option, it was just a complete lack of using the English language effectively on my part. My apologies.
I had meant to express that sentiment as: "this is harder to read than the `rpartition` option in my opinion, but if we're dropping that, then, this should be our solution". On re-read I... did not communicate that _whatsoever_. My apologies again, I did not mean to push you to spend time writing another alternative. 
I think the solutions are more or less equivalent (although I agree the value assignment should move elsewhere toward the top of the loop regardless, that way we would also be able to drop line 886 in config.py in this PR), although I personally am not a fan of logic statements if we can avoid them.
Either way, I'll leave it  to your discretion. Let me know which you prefer and I'll push (I have git stashes with all the discussed options so its trivial on my end)

Reply:
In this case, can we do https://github.com/spack/spack/pull/39279#discussion_r1310733035

and 

> I agree the value assignment should move elsewhere toward the top of the loop regardless

since we agree on that, let's move that initial assignment of `value` up

Reply:
Actually, on a second inspection, due to the nature of the loop, which is setting value to be a potentially nested dictionary, I think we can't set value to `value = syaml.load_config(components[-1])` as an alterative to `value = syaml.load_config(value)` as value may not be equal to `components[-1]` in some cases, unless I'm mistaken?

Reply:
I've made the other suggested change and pushed.

Reply:
> Actually, on a second inspection, due to the nature of the loop, which is setting value to be a potentially nested dictionary, I think we can't set value to value = syaml.load_config(components[-1])

That loop starts by setting `value = syaml.load_config(components[-1])`, and we need to set it outside of the loop as well, in 

```
if has_existing_value:
        value = syaml.load_config(value) # <---- right now, value is unset at this point, which I think is triggering the test failures
```